### PR TITLE
Allow Swagger UI CSS and JS paths to be configurable

### PR DESCRIPTION
**Pull request by mag1art**
Move CSS and JS to parameters, for available to use CDN or custom URL.

### CODE REVIEW

**Comment by martincostello** [comment]
Did you mean #2881?

**Comment by mag1art** [comment]
> Did you mean #2881?

Yes, this will add the ability to custom links to CSS and JS, and will also leave the current behavior with embedded (original) files by default.
My reason is to use cdn files.

File: src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIOptions.cs [hunk 1]
```diff
@@ -50,6 +50,21 @@ public class SwaggerUIOptions
...
+        /// </summary>
+        public string PathBundleJs { get; set; } = "./swagger-ui-bundle.js";
+
+        /// <summary>
+        /// Path to standalone js file.
+        /// </summary>
+        public string PathStandaloneJs { get; set; } = "./swagger-ui-standalone-preset.js";
+
+        /// <summary>
+        /// Path to css file.
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIOptions.cs:65]:
```suggestion
        /// Gets or sets the path or URL to the Swagger UI JavaScript bundle file.
        /// </summary>
        public string PathBundleJs { get; set; } = "./swagger-ui-bundle.js";

        /// <summary>
        /// Gets or sets the path or URL to the Swagger UI JavaScript standalone presets file.
        /// </summary>
        public string PathStandaloneJs { get; set; } = "./swagger-ui-standalone-preset.js";

        /// <summary>
        /// Gets or sets the path or URL to the Swagger UI CSS file.
```

File: src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIOptions.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIOptions.cs:1]: I think these properties would be better named something like:

- `ScriptBundlePath`
- `ScriptPresetsPath`
- `StylesPath`

**Review by martincostello — COMMENTED** [review]
Please add some tests that validate that the customised values are used in the rendered HTML.

File: src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIOptions.cs
Comment (mag1art) [reply src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIOptions.cs:1]: change names and add tests

**Comment by cremor** [comment]
> Resolves #2881.

I don't agree with that.

File: test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerUIIntegrationTests.cs [hunk 2]
```diff
@@ -88,5 +88,66 @@ public async Task SwaggerUIMiddleware_CanBeConfiguredMultipleTimes(string swagge
                 Assert.Contains(version, content);
             }
         }
+
+        [Theory]
+        [InlineData(typeof(Basic.Startup), "/index.html", "/swagger-ui.js", "/swagger-ui.css", "/swagger-ui-bundle.js", "/swagger-ui-standalone-preset.js")]
+        [InlineData(typeof(CustomUIConfig.Startup), "/swagger/index.html", "/swagger/swagger-ui.js", "/ext/custom-stylesheet.css", "/ext/custom-javascript.js", "/ext/custom-javascript.js")]
+        public async Task IndexUrl_ReturnsCustomScripts(
```
Comment (martincostello) [inline test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerUIIntegrationTests.cs:95]:
```suggestion
        public async Task IndexUrl_Returns_ExpectedAssetPaths(
```

File: test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerUIIntegrationTests.cs [hunk 3]
```diff
@@ -88,5 +88,66 @@ public async Task SwaggerUIMiddleware_CanBeConfiguredMultipleTimes(string swagge
...
+            Assert.Equal(HttpStatusCode.OK, jsResponse.StatusCode);
+
+            var cssResponse = await client.GetAsync(cssPath);
+            Assert.Equal(HttpStatusCode.OK, cssResponse.StatusCode);
+
+            var scriptBundleResponse = await client.GetAsync(scriptBundlePath);
+            Assert.Equal(HttpStatusCode.OK, scriptBundleResponse.StatusCode);
+
+            var scriptPresetsResponse = await client.GetAsync(scriptPresetsPath);
+            Assert.Equal(HttpStatusCode.OK, scriptPresetsResponse.StatusCode);
```
Comment (martincostello) [inline test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerUIIntegrationTests.cs:120]: Neither of these tests assert that you got the file you configured, just that a file happens to exist.

There's a missing assertion that the index content actually pointed to the files configured.

File: test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerUIIntegrationTests.cs [hunk 4]
```diff
@@ -88,5 +88,66 @@ public async Task SwaggerUIMiddleware_CanBeConfiguredMultipleTimes(string swagge
...
+        }
+
+        [Theory]
+        [InlineData(typeof(CustomUIConfig.Startup), "/swagger/index.html", "/swagger/swagger-ui.js", "/ext/custom-stylesheet-2.css", "/ext/custom-javascript-2.js", "/ext/custom-javascript-2.js")]
+        public async Task IndexUrl_ReturnsCustomScripts_NotFound(
```
Comment (martincostello) [inline test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerUIIntegrationTests.cs:125]: I don't think this test adds much value, it's essentially the same as the previous test except you tell the test to fetch files that don't exist and checks that fact - it doesn't really test anything about the configuration code.

**Comment by martincostello** [comment]
> I don't agree with that.

You're right, it doesn't - I was just fixing the fact that the PR originally said it fixed itself.

**Review by martincostello — APPROVED** [review]

**Comment by martincostello** [comment]
Thanks for your contribution - we'll look to merge this once we shipped 6.6.2 which has a bunch of bug fixes, rather than include new public API from this PR too (meaning it should then be 6.7.0).

File: README.md [hunk 5]
```diff
@@ -1254,6 +1255,19 @@ app.UseSwaggerUI(c =>
 }
 ```
 
+### Change CSS or JS Paths ###
+
+By default, the Swagger UI inclide default CSS and JS, but if you need to change path or URL (ex CDN):
```
Comment (martincostello) [inline README.md:1260]:
```suggestion
By default, the Swagger UI include default CSS and JS, but if you wish to change the path or URL (for example to use a CDN):
```

**Pull request merged**